caps/eth.md: add eth/63 protocol specification (#62)
ethereum/devp2p: caps/eth.md

# The eth/63 protocol

> 'eth' is protocol on the [RLPx](../rlpx.md) transport that facilitates exchange of

Reviewer: ```suggestion
'eth' is a protocol on the [RLPx](../rlpx.md) transport that facilitates exchange of
```

Ethereum blockchain information between peers. The current protocol version is **eth/63**.
See end of document for a list of changes in past protocol versions.

### Protocol Semantics

> Upon an active session, a Status message must be sent. Following the reception of the

Reviewer: I am not familiar with Eth protocol. So "Upon an active session" reads to me like "when a session becomes active", then a Status message must be sent. But the sentence following that says that the session is active only *after* some Status message is sent. This seems contradictory.

peer's Status message, the Ethereum session is active and any other messages may be sent.

All known transactions should be sent following the Status exchange with one or more
Transactions messages.

> Transactions messages should also be sent periodically as the node has new transactions to

Reviewer: ```suggestion
Transaction messages should also be sent periodically as the node has new transactions to
```

Author: This refers to the message 'Transactions'

> disseminate. A node should never send a transaction back to the peer that it can determine

Reviewer: ```suggestion
disseminate. A node should never send a transaction back to a peer that it can determine
```

already knows of it (either because it was previously sent or because it was informed from
this peer originally).

New blocks are propagated using the NewBlock and NewBlockHashes messages. The NewBlock
message includes the full block and is sent to a small fraction of connected peers
(usually sqrt(peers)). All other peers are sent a NewBlockHashes message containing just
the hash of the new block. Those peers may request the full block later if they fail to

> receive it from anyone within reasonable time.

Reviewer: ```suggestion
receive it from anyone within a reasonable time.
```

Author: Would like to leave this one as is

> Blocks are typically re-propagated to all connected peers as soon as basic validity of the

Reviewer: To all? Or sqrt(all)?

Author: Resolved by moving the sentence around.

announcement has been established (e.g. after the proof-of-work check).

### Basic Chain Synchronization

> Two peers connect and send their Status message. Status includes the Total Difficulty(TD)

Reviewer: "Connect" in "two peers connect" needs disambiguation/ clarification. I think what you mean here is that the underlying RLPx transport establishes a connection.

and hash of their best block.

The client with the worst TD then proceeds to download block headers using the
BlockHeadersFromNumber message. It verifies proof-of-work values in received headers and
fetches block bodies using the GetBlockBodies message. Received blocks are executed using

> the Ethereum Virtual Machine, recreating the state tree.

Reviewer: is it tree? or trie? I think trie right?

Author: Tries are special trees. Now changed to tree everywhere.

Note that header downloads, block body downloads and block execution may happen
concurrently.

### State Synchronization (a.k.a. "fast sync")

eth/63 also allows synchronizing transaction execution results ("state"). This may be
faster than re-executing all transactions but comes at the expense of some security.

State synchronization typically proceeds by downloading the chain of block headers,
verifying their proof-of-work values. Block bodies are requested as in the Basic Chain
Synchronization section but block transactions aren't executed. Instead, the client picks
a block near the head of the chain and downloads merkle tree nodes and contract code
incrementally by requesting the root node, it's children, grandchildren, ... until the
entire tree is synchronized.

## Protocol Messages

### Status (0x00)

`[protocolVersion: P, networkId: P, td: P, bestHash: B_32, genesisHash: B_32, number: P]`

Inform a peer of its current ethereum state. This message should be sent _after_ the

> initial handshake and _prior_ to any **ethereum** related messages.

Reviewer: Again I think if the 'initial handshake' is referring to the RLPx transport handshake, it should not really be part of this protocol. Or rather, I do remember being confused reading the original document a year ago, trying to establish if this handshake they were referring to was part of this protocol or some other. Maybe this could be rephrased to say "after the underlying transport has completed the handshake and the connection is established, the first message of this Eth protocol must be...."

Author: I don't want to refer to RLPx specifics in this document.

* `protocolVersion`: the current protocol version, 63

> * `networkId`: Integer identifying the blockchain, see table below

Reviewer: Not super important but I have noticed that even amongst battle-hardened Ethereum experts there is still confusion about the distinction between Network ID and Chain ID. It might be worth adding a helpful note here.

| 3  | Ropsten (current PoW testnet)      |
| 4  | [Rinkeby](https://www.rinkeby.io/) |

### NewBlockHashes (0x01)

`[[hash_0: B_32, number_0: P], [hash_1: B_32, number_1: P], ...]`

Specify one or more new blocks which have appeared on the network. To be maximally
helpful, nodes should inform peers of all blocks that they may not be aware of. Including
hashes that the sending peer could reasonably be considered to know (due to the fact they
were previously informed of because that node has itself advertised knowledge of the
hashes through NewBlockHashes) is considered bad form, and may reduce the reputation of
the sending node. Including hashes that the sending node later refuses to honour with a
proceeding GetBlockHeaders message is considered bad form, and may reduce the reputation
of the sending node.

### Transactions (0x02)

`[[nonce: P, receivingAddress: B_20, value: P, ...], ...]`

Specify (a) transaction(s) that the peer should make sure is included on its transaction
queue. The items in the list (following the first item `0x12`) are transactions in the
format described in the main Ethereum specification. Nodes must not resend the same

> transaction to a peer in the same session. This packet must contain at least one (new)

Reviewer: Nodes must also not send a transaction back to that same peer if it was learned from that peer?

transaction.

### GetBlockHeaders (0x03)

`[block: {P, B_32}, maxHeaders: P, skip: P, reverse: P in {0, 1}]`

Require peer to return a BlockHeaders message. Reply must contain a number of block
headers, of rising number when `reverse` is `0`, falling when `1`, `skip` blocks apart,
beginning at block `block` (denoted by either number or hash) in the canonical chain, and
with at most `maxHeaders` items.

### BlockHeaders (0x04)

`[blockHeader_0, blockHeader_1, ...]`

Reply to GetBlockHeaders. The items in the list (following the message ID) are block
headers in the format described in the main Ethereum specification, previously asked for
in a GetBlockHeaders message. This may validly contain no block headers if no block
headers were able to be returned for the GetBlockHeaders query.

### GetBlockBodies (0x05)

`[hash_0: B_32, hash_1: B_32, ...]`

> Require peer to return a BlockBodies message. Specify the set of blocks that we're

Reviewer: Is there a max number of hashes we can send?

interested in with the hashes.

### BlockBodies (0x06)

`[[transactions_0, uncles_0] , ...]`

Reply to GetBlockBodies. The items in the list are some of the blocks, minus the header,
in the format described in the main Ethereum specification, previously asked for in a

> GetBlockBodies message. This may validly contain no items if no blocks were able to be

Reviewer: ```suggestion
GetBlockBodies message. This may be empty if no blocks were able to be
```

returned for the GetBlockBodies query.

### NewBlock (0x07)

`[[blockHeader, transactionList, uncleList], totalDifficulty]`

Specify a single block that the peer should know about. The composite item in the list
(following the message ID) is a block in the format described in the main Ethereum
specification.

- `totalDifficulty` is the total difficulty of the block (aka score).

### BlockHashesFromNumber (0x08)

`[number: P, maxBlocks: P]`

Requires peer to reply with a BlockHashes message. Message should contain block with

> that of number `number` on the canonical chain. Should also be followed by subsequent

Reviewer: "Message should contain block with that of number..." not sure what that is supposed to mean.

blocks, on the same chain, detailing a number of the first block hash and a total of
hashes to be sent. Returned hash list must be ordered by block number in ascending order.

### GetNodeData (0x0d)

`[hash_0: B_32, hash_1: B_32, ...]`

Require peer to return a NodeData message containing state trie nodes or contract code
matching the requested hashes.

### NodeData (0x0e)

`[value_0: B, value_1: B, ...]`

Provide a set of state trie nodes or contract code blobs which correspond to previously
asked hashes from GetNodeData. Does not need to contain all; best effort is fine. If it

> contains none, then has no information for previous GetNodeData hashes.

Reviewer: ```suggestion
contains none, then it has no relevant information on the requested GetNodeData hashes.
```

### GetReceipts (0x0f)

`[blockHash_0: B_32, blockHash_1: B_32, ...]`

Require peer to return a Receipts message containing the receipts of the given block
hashes.

### Receipts (0x10)

`[[receipt_0, receipt_1], ...]`

> Provide a set of receipts which correspond to previously asked in GetReceipts.

Reviewer: ```suggestion
Provide a set of receipts which correspond to those previously requested in GetReceipts.
```
